File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.py
import heapq

def solution(cards):
    answer = 0
    que = []
    
    n = len(cards)
    visited = [False] * n
    
    for i in range(1, n+1):
        if visited[i-1]:
            continue
            
        g = set()
        k = i
        while True:
            g.add(cards[k-1])
            visited[k-1] = True
            k = cards[k-1]
            if visited[k-1]:
                break
        # print(g)
        heapq.heappush(que, -len(g))
    if len(que) > 1:
        return heapq.heappop(que) * heapq.heappop(que)
    else:
        return 0

File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.kt
import java.util.*
class Solution {
    fun solution(cards: IntArray): Int {
        var answer: Int = 0

        val visited = BooleanArray(cards.size)
        val queue = PriorityQueue<Int>()
        for(i in cards.indices) {
            if(visited[i]){
                continue
            }
            visited[i] = true

            var count = 1
            var next = cards[i]
            while(!visited[next - 1]) {
                visited[next - 1] = true
                next = cards[next - 1]
                count++
            }
            queue.add(-count)

        }

        if(queue.size >= 2){
            return queue.poll() * queue.poll()
        }
        return answer
    }
}